counter.go

```diff
@@ -0,0 +1,127 @@
+package signalfx
+
+import (
+	"math"
+	"sync/atomic"
+	"time"
+)
+
+// A Counter represents a Counter metric (i.e., its internal value is
+// reset to zero by its PostReportHook, although this should not be
+// meaningful to client code).
+type Counter struct {
+	metric     string
+	dimensions map[string]string
+	value      uint64
+}
+
+// NewCounter returns a new Counter with the specified parameters.  It
+// does not copy the dimensions; client code should take care not to
+// modify them in a goroutine-unsafe manner.
+func NewCounter(
+	metric string,
+	dimensions map[string]string,
+	value uint64,
+) *Counter {
+	return &Counter{metric: metric, dimensions: dimensions, value: value}
+}
+
+// Inc increments a Counter's internal state.  It is goroutine-safe.
+func (c *Counter) Inc(delta uint64) uint64 {
+	return atomic.AddUint64(&c.value, delta)
+}
+
+// DataPoint returns a DataPoint reflecting the Counter's internal
+// state.  It does not reset that state, leaving that to
+// PostReportHook.
+func (c *Counter) DataPoint() *DataPoint {
+	value := atomic.LoadUint64(&c.value)
+	if value == 0 || value > math.MaxInt64 {
+		return nil
+	}
+	return &DataPoint{
+		Metric:     c.metric,
+		Timestamp:  time.Now(),
+		Type:       CounterType,
+		Dimensions: c.dimensions,
+		Value:      int64(value),
+	}
+}
+
+// PostReportHook resets a Counter's internal state.  It does this
+// intelligently, by subtracting the successfully-reported value from
+// the state (this way, the Counter may have been incremented in
+// between the create of the reported DataPoint and the reset, and the
+// state will be consistent).  Calling PostReportHook with a negative
+// value will result in a panic: counters may never take negative
+// values.
+func (c *Counter) PostReportHook(v int64) {
+	if v < 0 {
+		panic("negative counter should be impossible")
```

Comment:
It is definitely ugly, but what's the alternative? That's an honest question. I only see two:

 - Just do nothing if Reporter.Report tells a counter that it just reported a negative value. This has the virtue of not crashing a production system. But if this occurs, that production system's memory is hosed, in which case it's untrustworthy and should crash (and be replaced…) anyway.
 - Implement a very complex error-recovery process where the definition of 'recovery' is punted in some way all the way up to the code which calls Reporter.RunInBackground. This has the virtue of enabling some sort of recovery (maybe the main program could choose to remove or recreate the counter or cumulative counter?), although since it could only be the result of memory corruption, the 'recovered' process is likely to crash again soon.

I think that the key point with this code is that it is detecting a memory corruption error; it's the equivalent of `if (*a != *a) …` in unthreaded C. There is absolutely no way that `Counter.DataPoint`, 'WrappedCounter.DataPoint`, `CumulativeCounter.DataPoint` or `WrappedCumulativeCounter.DataPoint` will ever produce a negative `DataPoint.Value`, and thus if `Counter.PostReportHook`, `WrappedCounter.PostReportHook`, `CumulativeCounter.PostReportHook` or `WrappedCumulativeCounter.PostReportHook` ever receives a value less than zero, then it's an exceptionally grave runtime error.

If traces the value, it is produced by one of the `DataPoint` methods of a struct, recorded in a `struct{HookedMetric,int64}` by `Reporter.Report`, then handed back to the original type's `PostReportHook`.

I guess another way of looking at it is that this is the moral equivalent of a slice bounds violation or nil pointer access: when it happens, the system just can't be trusted to recover cleanly.

I guess I could just remove the check, on the grounds that `if 2 == 1` will never happen and the cost of maintaining the check is higher than it could be worth…